helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -581,6 +585,46 @@ public int compare(String ins1, String ins2) {
       Integer p1 = Integer.MAX_VALUE;
       Integer p2 = Integer.MAX_VALUE;
 
+      // Order by MZ representation (overrepresented should be dropped first)
+      String mz1 = null;
+      String mz2 = null;
+      String faultZoneType = null;
+
+      if (_cache != null && _cache.getInstanceConfigMap() != null &&
+          _cache.getClusterConfig() != null && _cache.getClusterConfig().isTopologyAwareEnabled()) {
+
+        faultZoneType = _cache.getClusterConfig().getFaultZoneType();
+        mz1 = _cache.getInstanceConfigMap().get(ins1) != null ?
+            _cache.getInstanceConfigMap().get(ins1).getDomainAsMap().get(faultZoneType) : null;
+        mz2 = _cache.getInstanceConfigMap().get(ins2) != null ?
+            _cache.getInstanceConfigMap().get(ins2).getDomainAsMap().get(faultZoneType) : null;
+      }
+
+      // If mz is not valid for either instance, skip mz comparison
+      // If instances are in the same MZ, skip mz comparison
+      if (mz1 != null && mz2 != null && !mz1.equals(mz2)) {
+        // Count number of replicas in each MZ
+        int mz1Count = 0;
+        int mz2Count = 0;
+        for (String instance : _currentStateMap.keySet()) {
+            if (ins1.equals(instance) || ins2.equals(instance)) {
+              continue;
+            }
+
+            String currMz = _cache.getInstanceConfigMap().get(instance).getDomainAsMap().get(faultZoneType);
+
+            if (mz1.equals(currMz)) {
+              mz1Count++;
+            } else if (mz2.equals(currMz)) {
+              mz2Count++;
+            }
+        }
+        if (mz1Count != mz2Count) {
+          return mz1Count - mz2Count;
+        }
```

Comment:
You are counting instances per mz. It is expensive to recompute the counting looping through the current state map.

You can make a compute and loop current state to compute MZ counting mapping in constructor once. And every compare can leverage that map.

Reply:
Good points on both issues, thank you for the review. I'll address this